### PR TITLE
Fixes for translation fallbacks and JS catalog loading (SH-17, SH-90)

### PR DESCRIPTION
**Pull request by tulimaki**
* MultiLanguageModelForm: Use Parler default as a default
* MultiLanguageModelForm: Avoid partially/empty translation objects
* Core: Fix bug in product and category slugs generation
* Admin: Add key prefix to JavaScript catalog page cache
* Admin: Add lang parameter for JS catalog load
* Xtheme: Skip adding JS-catalog for editing
* Tests: Add test for loading JS catalog

### CODE REVIEW

File: shuup/utils/multilanguage_model_form.py [hunk 1]
```diff
@@ -26,14 +27,18 @@ def to_language_codes(languages):
     return languages
 
 
+def get_default_language():
+    return getattr(settings, "PARLER_DEFAULT_LANGUAGE_CODE", get_language())
```
Comment (Pikkupomo) [inline shuup/utils/multilanguage_model_form.py]: Should we even enforce this value? (`settings.PARLER_DEFAULT_LANGUAGE_CODE`) without the code default is again all over the place, right?

File: doc/changelog.rst [hunk 2]
```diff
@@ -60,6 +68,12 @@ Guide
 General/miscellaneous
 ~~~~~~~~~~~~~~~~~~~~~
 
+- MultiLanguageModelForm: Avoid partially/empty translation objects
+  - Delete untranslated objects from database
+  - Only set translation object to database if it is translated
+  - Ensure required fields if language is partially translated
```
Comment (Pikkupomo) [inline doc/changelog.rst]: These won't render properly when viewing the file in html. I think you have to add more spaces or something

File: shuup/utils/multilanguage_model_form.py [hunk 3]
```diff
@@ -54,6 +55,7 @@ def __init__(self, **kwargs):
         self.trans_field_map = defaultdict(dict)
         self.trans_name_map = defaultdict(dict)
         self.translated_field_names = []
+        self.required_if_translated_field_names = []
```
Comment (gurch101) [inline shuup/utils/multilanguage_model_form.py]: can't this just be named `required_translated_field_names`? Also minor but if order doesn't matter, this should be a set.

File: shuup/utils/multilanguage_model_form.py [hunk 4]
```diff
@@ -95,18 +99,38 @@ def __getitem__(self, key):
...
             (trans.language_code, trans)
             for trans
-            in translations_model.objects.filter(master=instance, language_code__in=self.languages)
+            in translations_model.objects.filter(master_id=instance.id, language_code__in=self.languages)
         )
         for lang, field_map in six.iteritems(self.trans_field_map):
+            translation_fields = dict((src_name, data.get(src_name)) for src_name in field_map)
+            translation = current_translations.get(lang)
+            # Add translation only if at least one translated field is given
+            if not [value for value in translation_fields.values() if value is not None and value != ""]:
```
Comment (gurch101) [inline shuup/utils/multilanguage_model_form.py]: could this be
```
if not any(translation_fields.values()):
```

**Comment by tulimaki** [comment]
Migration conflicts with https://github.com/shuup/shuup/pull/780 so be carefull when merging. I think I can change this when the language PR is merged.

File: shuup/admin/templates/shuup/admin/macros/multilanguage.jinja [hunk 5]
```diff
@@ -40,6 +40,9 @@ When using multiple language_dependent_content_tabs in a single page, do remembe
                 {% endif %}
             </div>
         {% endfor %}
+        <p class="text-center">
+            {% trans %}Partially translated languages does not fallback to default language{% endtrans %}
```
Comment (gurch101) [inline shuup/admin/templates/shuup/admin/macros/multilanguage.jinja]: Are we sure we want this message to always be visible? Maybe this is something that should be documented in the guide.
If we do then maybe use this: "Partially translated fields do not fall back to the default language"

File: shuup/core/models/_categories.py [hunk 6]
```diff
@@ -136,10 +136,10 @@ def is_visible(self, customer):
         return True
 
     @staticmethod
-    def _get_slug_name(self):
+    def _get_slug_name(self, translation):
```
Comment (gurch101) [inline shuup/core/models/_categories.py]: does this work? `self.safe_translation_getter('title', language_code=language_code)`
If you pass the `language_code` to this function then `generate_multilanguage_slugs` could be made much simpler

File: shuup/utils/multilanguage_model_form.py [hunk 7]
```diff
@@ -125,15 +151,27 @@ def _save_translation(self, instance, translation):
...
+        # slug from translated name.
+        self._save_master(commit)
         return self.instance
 
     def _set_fields_for_language(self, language):
-        self.instance.set_current_language(language)
-        for field in self.translation_fields:
-            value = self.cleaned_data["%s__%s" % (field.name, language)]
-            field.save_form_data(self.instance, value)
+        original_language = self.instance.get_current_language()
```
Comment (gurch101) [inline shuup/utils/multilanguage_model_form.py]: sorry for the late comment :) Looking at the django-parler docs, we should be able to use a context manager here:
```
from parler.utils.context import switch_language

with switch_language(self.instance, language):
    do stuff
```

File: shuup/core/models/_categories.py (hunk 6 above)
Comment (tulimaki) [inline shuup/core/models/_categories.py]: The reason we are passing translation object here is that the product/category-object might not contain the translation. Which will lead in the situation that the `safe_translation_getter` does not return right `title` for the object.

At least the `shuup.utils.multlanguage_model_form:_save_translations` causes situation that the object that we are generating slug for does not have translation even if the translation object is already in the database. Not sure if that is correct either but probably we don't want to change this functionality. @Pikkupomo what do you think about this?

**Pull request merged**